src/app/ReadHandler.cpp

```diff
@@ -31,13 +31,28 @@
 #include <lib/core/TLVUtilities.h>
 #include <messaging/ExchangeContext.h>
 
+#include <app-common/zap-generated/attributes/Accessors.h>
 #include <app/ReadHandler.h>
 #include <app/reporting/Engine.h>
 
+#if CHIP_CONFIG_ENABLE_ICD_SERVER
+#include <app/icd/IcdManagementServer.h> // nogncheck
```

Comment:
The nogncheck here indicates we have circular dependencies, right?  Would it not be better to acknowledge that in the gn files instead of skipping the checks?

And as a followup, perhaps we should break up src/app so that we don't have these dependency cycles...

Reply:
Yes, and yes. I have spend a lot of time this week looking for a way to square this circle, and I none of the walk-arounds I found is satisfactory.
The intent of this PR is to add a single missing condition to the logic, but that condition requires getting the value of an attribute from the Interaction layer. Until such day we split src/app, as you said, these ugly patches will be hard to avoid.

Reply:
@rcasallas-silabs I guess my question was this: why are we not using `allow_circular_includes_from` in the gn files as our hack instead of these "disable the check" bits?  Did it not work?  It would be preferable because easier to search for and only one spot.

Reply:
@bzbarsky-apple There is something more than just the circular dependencies here. the gn check doesn't deal with preprocessor defines and will ignore the condition include on`#if CHIP_CONFIG_ENABLE_ICD_SERVER`.

It will do the check as if `<app/icd/IcdManagementServer.h> `is always included. Depending on the gn files the needed dependency is probably not added by default (only added in the case `chip_enable_icd_server` arg is true)

Reply:
> It will do the check as if <app/icd/IcdManagementServer.h> is always included.

If that's what's going on, then we should clearly document that at the place where we use nogncheck, so it's clear why it's there.

Reply:
Yeah we should add some comment.

Here is a section of the gn doc that explains it
```  
  GN's header checker helps validate that the includes match the build
  dependency graph. Sometimes an include might be conditional or otherwise
  problematic, but you want to specifically allow it. In this case, it can be
  whitelisted.

  Include lines containing the substring "nogncheck" will be excluded from
  header checking. The most common case is a conditional include:

    #if defined(ENABLE_DOOM_MELON)
    #include "tools/doom_melon/doom_melon.h"  // nogncheck
    #endif

  If the build file has a conditional dependency on the corresponding target
  that matches the conditional include, everything will always link correctly:

    source_set("mytarget") {
      ...
      if (enable_doom_melon) {
        defines = [ "ENABLE_DOOM_MELON" ]
        deps += [ "//tools/doom_melon" ]
      }

  But GN's header checker does not understand preprocessor directives, won't
  know it matches the build dependencies, and will flag this include as
  incorrect when the condition is false.
```

Reply:
I have removed the problematic dependencies, at the cost of ignoring the ZAP configuration.